database/gorm/query.go

```diff
@@ -780,6 +866,82 @@ func (r *QueryImpl) Where(query any, args ...any) ormcontract.Query {
 	return r.setConditions(conditions)
 }
 
+func (r *QueryImpl) WhereHas(relation string, callback func(query ormcontract.Query) (ormcontract.Query, error), args ...any) ormcontract.Query {
+
+	const (
+		gt  = ">"
+		gte = ">="
+		lte = ">="
+		lt  = "<"
+	)
+
+	var (
+		op    string
+		count int64
+	)
+
+	if argsLen := len(args); argsLen > 0 && argsLen != 1 {
+		o, ok := args[0].(string)
+		if !ok {
+			return r.setError(errors.New("the first argument should be string, it uses as operator"))
+		}
+
+		c, err := cast.ToInt64E(args[1])
+		if err != nil {
+			return r.setError(errors.New("the second argument should be int64, it uses as count"))
+		}
+
+		op = o
+		count = c
+	}
+
+	mSchema, err := getModelSchema(r.conditions.model, r.instance)
+	if err != nil {
+		return r.setError(fmt.Errorf("faild to get model schema, the model should be set before using this method. %w", err))
+	}
+
+	modifiedQuery, err := callback(r.new(r.instance))
+	if err != nil {
+		return r.setError(err)
+	}
+	rel, ok := mSchema.Relationships.Relations[relation]
+	if !ok {
+		r.setError(errors.New("the relation not found"))
+	}
+	relModel := getZeroValueFromReflectType(rel.Field.FieldType)
+
+	fk := rel.References[0].ForeignKey.DBName
+	ft := rel.FieldSchema.Table
+	table := mSchema.Table
+
+	switch rel.Type {
+	case schema.BelongsTo:
+		pk := rel.FieldSchema.PrioritizedPrimaryField.DBName
+		modifiedQuery = modifiedQuery.Where(database.QuoteConcat(ft, pk) + " = " + database.QuoteConcat(table, fk))
+	case schema.HasOne, schema.HasMany:
+		pk := mSchema.PrioritizedPrimaryField.DBName
+		modifiedQuery = modifiedQuery.Where(database.QuoteConcat(ft, fk) + " = " + database.QuoteConcat(table, pk))
+	case schema.Many2Many:
+		joinTable := rel.JoinTable.Table
+		pk := mSchema.PrioritizedPrimaryField.DBName
+		modifiedQuery = modifiedQuery.
+			Join("inner join " +
+				database.Quote(joinTable) +
+				" on " +
+				database.QuoteConcat(mSchema.Table, pk) +
+				" = " + database.QuoteConcat(joinTable, fk))
+	}
+	needCountQuery := !((count == 0 && slices.Contains([]string{lt, lte, gt, gte}, op)) || op == "")
+	if !needCountQuery {
+		modifiedQueryImpl := modifiedQuery.Model(relModel).(*QueryImpl)
+		return r.Where("EXISTS (?)", modifiedQueryImpl.buildConditions().instance)
+	}
+
+	modifiedQueryImpl := modifiedQuery.Model(relModel).Select("count(*)").(*QueryImpl)
+	return r.Where("(?) "+op+" ?", modifiedQueryImpl.buildConditions().instance, count)
+
+}
```

Comment:
Do you test the logic locally? Is there a screenshot, please? You can add it in the PR description.